src/software/amazon/ionschema/internal/TypeReference.kt

```diff
@@ -0,0 +1,98 @@
+package software.amazon.ionschema.internal
+
+import software.amazon.ion.IonStruct
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonText
+import software.amazon.ion.IonValue
+import software.amazon.ionschema.InvalidSchemaException
+import software.amazon.ionschema.Schema
+
+/**
+ * Provides a factory method that translates an ISL type reference into a function
+ * that returns a Type instance.
+ *
+ * Types that can't be resolved yet are instantiated as [TypeReferenceDeferred] objects
+ * that are resolved by [SchemaImpl.resolveDeferredTypeReferences] prior to asserting
+ * that the schema is valid.
+ */
+internal class TypeReference private constructor() {
+    companion object {
+        fun create(ion: IonValue, schema: Schema, isField: Boolean = false): () -> TypeInternal {
+            var type: TypeInternal? = null
+            if (!ion.isNullValue) {
+                type = when (ion) {
+                    is IonStruct -> {
```

Comment:
suggestion: avoid inlining a lot of logic in when statements as it can get hard to read pretty fast. Dispatch to a private fun, example: 

```kotlin 
// very clear that the non null Ion can only be a struct or symbol
type = when(ion) {
    is IonStruct -> handleStruct(...)
    is IonSymbol -> handleSymbol(...)
    else -> throw InvalidSchemaException("Unable to resolve type reference '$ion'")
   }
```